Update ANCM error pages and diagnostics (#12605)
aspnet/AspNetCore.Docs: aspnetcore/host-and-deploy/iis/troubleshoot.md

::: moniker-end

::: moniker range="= aspnetcore-3.0"

> ### 500.30 In-Process Startup Failure

Reviewer: Suggest moving this one (the generic failure, right?) to the end. That way the flow is basically: All the errors that are well-known and understood, then at the end a section that says "If something s p o o k y happens, you get this error. Check the event log"

The worker process fails. The app doesn't start.

The ASP.NET Core Module attempts to start the .NET Core CLR in-process, but it fails to start. The cause of a process startup failure can usually be determined from entries in the [Application Event Log](#application-event-log) and the [ASP.NET Core Module stdout log](#aspnet-core-module-stdout-log).

### 500.31 ANCM Failed to Find Native Dependencies

The worker process fails. The app doesn't start.

> The ASP.NET Core Module attempts to start the .NET Core CLR in-process, but it fails to start. The cause of this startup failure most likely due to the version of `Microsoft.NETCore.App` or `Microsoft.AspNetCore.App` not being found. For example, if the application is deployed to target 3.0 and that version doesn't exist on the machine, this error can occur. The specific error will look like the following:

Reviewer: ```suggestion
The ASP.NET Core Module attempts to start the .NET Core runtime in-process, but it fails to start. The most common cause of this startup failure is when the `Microsoft.NETCore.App` or `Microsoft.AspNetCore.App` runtime is not installed. For example, if the application is deployed to target ASP.NET Core 3.0 and that version doesn't exist on the machine, this error can occur. An example error message is shown below:
```

```
The specified framework 'Microsoft.NETCore.App', version '3.0.0' was not found.
  - The following frameworks were found:
      2.2.1 at [C:\Program Files\dotnet\x64\shared\Microsoft.NETCore.App]
      3.0.0-preview5-27626-15 at [C:\Program Files\dotnet\x64\shared\Microsoft.NETCore.App]
      3.0.0-preview6-27713-13 at [C:\Program Files\dotnet\x64\shared\Microsoft.NETCore.App]
      3.0.0-preview6-27714-15 at [C:\Program Files\dotnet\x64\shared\Microsoft.NETCore.App]
      3.0.0-preview6-27723-08 at [C:\Program Files\dotnet\x64\shared\Microsoft.NETCore.App]
```

> To fix this error, either:

Reviewer: ```suggestion
The error message lists all the installed versions, as well as the version requested by the application. To fix this error, either:
```

> * Install the appropriate version of dotnet on the machine.

Reviewer: ```suggestion
* Install the appropriate version of .NET Core on the machine.
```

> * Target the correct version of dotnet.

Reviewer: ```suggestion
* Change the app to target a version of .NET Core that is present on the machine.
```

> * Publish the application as Standalone.

Reviewer: ```suggestion
* Publish the app as a [self-contained deployment](/dotnet/core/deploying/#self-contained-deployments-scd).
```

> When running in development (ASPNETCORE_ENVIRONMENT is set to Development), the specific error that occured will be written to the HTTP response. The cause of a process startup failure can also be found in the [Application Event Log](#application-event-log).

Reviewer: ```suggestion
When running in development (the `ASPNETCORE_ENVIRONMENT` environment variable is set to Development), the specific error that occurred will be written to the HTTP response. The cause of a process startup failure can also be found in the [Application Event Log](#application-event-log).
```

### 500.32 ANCM Failed to Load dll

The worker process fails. The app doesn't start.

> The application was likely published for a different bitness than w3wp.exe/iisexpress.exe is running as. For example, if the worker process is running as a 32-bit application and the application was published to target 64-bit, this error will occur. The first dll that will hit this error will be `hostfxr.dll`, which is responsible for starting the dotnet application.

Reviewer: ```suggestion
The most common cause for this is that the app was published for an incompatible processor architecture. For example, if the worker process is running as a 32-bit application and the application was published to target 64-bit, this error will occur.
```

Re `The first dll that will hit this error will be `hostfxr.dll`, which is responsible for starting the dotnet application.`. Is that relevant to the user? Do they see that name in the error anywhere?

To fix this error, either:

> * Publish the application for the same bitness as the worker process

Reviewer: ```suggestion
* Re-publish the app for the same processor architecture as the worker process.
```

> * Publish the application as portable.

Reviewer: ```suggestion
* Publish the app as a [framework-dependent deployment](/dotnet/core/deploying/#framework-dependent-executables-fde).
```

### 500.33 ANCM Request Handler Load Failure

The worker process fails. The app doesn't start.

> The application didn't reference a package or shared framework that included the native handler used by ANCM. 

Reviewer: ```suggestion
The app didn't reference the `Microsoft.AspNetCore.App` framework. Only apps targeting that framework can be hosted by ANCM.
```

> To fix this error, make sure Microsoft.AspNetCore.App is referenced by the application.

Reviewer: ```suggestion
To fix this error, make sure the app is targeting the `Microsoft.AspNetCore.App` framework. Check the `.runtimeconfig.json` to verify the framework targeted by the app.
```

### 500.34 ANCM Mixed Hosting Models Not Supported

> The worker process cannot run both an inprocess and out of process application in the same worker process.

Reviewer: ```suggestion
The worker process cannot run both an in-process and out-of-process application in the same worker process.
```

> To fix this error, please target a different application pool for the second application.

Reviewer: ```suggestion
To fix this error, run apps in separate IIS application pools.
```

### 500.35 ANCM Multiple In-Process Applications in same Process

The worker process cannot run two inprocess applications in the same worker process.

> To fix this error, please target a different application pool for the second application.

Reviewer: ```suggestion
To fix this error, run apps in separate IIS application pools.

```

### 500.36 ANCM Out-Of-Process Handler Load Failure

> The out of process request handler, aspnetcorev2_outofprocess.dll, could not be found next to the aspnetcorev2.dll.

Reviewer: ```suggestion
The out-of-process request handler, `aspnetcorev2_outofprocess.dll`, could not be found next to the `aspnetcorev2.dll`. This indicates a corrupted installation of the ASP.NET Core Module.
```

> To fix this error, please confirm that the dll aspnetcorev2_outofprocess.dll in a directory with a version number next to aspnetcorev2.dll

Reviewer: ```suggestion
To fix this error, repair your installation of the Windows Hosting Bundle (for IIS) or Visual Studio (for IIS Express).
```

### 500.37 ANCM Failed to Start Within Startup Time Limit

ANCM failed to start within the provied startup time limit. By default, the timeout is 120 seconds.

> This can occur if multiple inprocess applications are being started at the same time. 

Reviewer: ```suggestion
This can occur when starting a large number of apps on the same machine. Check for CPU/Memory usage spikes on the server during startup. You may need to stagger the startup process of multiple apps.
```

### 500.0 In-Process Handler Load Failure

The worker process fails. The app doesn't start.

> The cause of a process startup failure can also be found in the [Application Event Log](#application-event-log).

Reviewer: Is this a handler load failure or an app load failure? Can the user be expected to do anything here? If not, let's bounce them to support (and/or to GitHub) at this point:

```suggestion
An unknown error occurred loading ANCM components. Contact [Microsoft Support](https://support.microsoft.com/oas/default.aspx?prid=15832) (select "Developer Tools" then "ASP.NET Core") for assistance, ask on Stack Overflow, or file an issue on our [GitHub repository](https://github.com/aspnet/AspNetCore).
```

::: moniker-end

> ::: moniker range="= aspnetcore-3.0"

Reviewer 2: Should this be >= 3.0?

```suggestion
::: moniker range=">= aspnetcore-3.0"
```

### 500.31 ANCM Failed to Find Native Dependencies

The worker process fails. The app doesn't start.

> The ASP.NET Core Module attempts to start the .NET Core runtime in-process, but it fails to start. The most common cause of this startup failure is when the `Microsoft.NETCore.App` or `Microsoft.AspNetCore.App` runtime is not installed. For example, if the application is deployed to target ASP.NET Core 3.0 and that version doesn't exist on the machine, this error can occur. An example error message is shown below:

Reviewer 2: ```suggestion
The ASP.NET Core Module attempts to start the .NET Core runtime in-process, but it fails to start. The most common cause of this startup failure is when the `Microsoft.NETCore.App` or `Microsoft.AspNetCore.App` runtime isn't installed. If the app is deployed to target ASP.NET Core 3.0 and that version doesn't exist on the machine, this error occurs. An example error message follows:
```

```
The specified framework 'Microsoft.NETCore.App', version '3.0.0' was not found.
  - The following frameworks were found:
      2.2.1 at [C:\Program Files\dotnet\x64\shared\Microsoft.NETCore.App]
      3.0.0-preview5-27626-15 at [C:\Program Files\dotnet\x64\shared\Microsoft.NETCore.App]
      3.0.0-preview6-27713-13 at [C:\Program Files\dotnet\x64\shared\Microsoft.NETCore.App]
      3.0.0-preview6-27714-15 at [C:\Program Files\dotnet\x64\shared\Microsoft.NETCore.App]
      3.0.0-preview6-27723-08 at [C:\Program Files\dotnet\x64\shared\Microsoft.NETCore.App]
```

> The error message lists all the installed versions, as well as the version requested by the application. To fix this error, either:

Reviewer 2: ```suggestion
The error message lists all the installed .NET Core versions and the version requested by the app. To fix this error, either:

```

* Install the appropriate version of .NET Core on the machine.

> * Change the app to target a version of .NET Core that is present on the machine.

Reviewer 2: ```suggestion
* Change the app to target a version of .NET Core that's present on the machine.
```

* Publish the app as a [self-contained deployment](/dotnet/core/deploying/#self-contained-deployments-scd).

> When running in development (the `ASPNETCORE_ENVIRONMENT` environment variable is set to Development), the specific error that occurred will be written to the HTTP response. The cause of a process startup failure can also be found in the [Application Event Log](#application-event-log).

Reviewer 2: ```suggestion
When running in development (the `ASPNETCORE_ENVIRONMENT` environment variable is set to `Development`), the specific error is written to the HTTP response. The cause of a process startup failure is also found in the [Application Event Log](#application-event-log).
```

### 500.32 ANCM Failed to Load dll

The worker process fails. The app doesn't start.

> The most common cause for this is that the app was published for an incompatible processor architecture. For example, if the worker process is running as a 32-bit application and the application was published to target 64-bit, this error will occur.

Reviewer 2: ```suggestion
The most common cause for this error is that the app is published for an incompatible processor architecture. If the worker process is running as a 32-bit app and the app was published to target 64-bit, this error occurs.
```

> To fix this error, either:

Reviewer 2: ```suggestion
To fix this error, either:

```

> * Re-publish the app for the same processor architecture as the worker process.

Reviewer 2: ```suggestion
* Republish the app for the same processor architecture as the worker process.
```

* Publish the app as a [framework-dependent deployment](/dotnet/core/deploying/#framework-dependent-executables-fde).

### 500.33 ANCM Request Handler Load Failure

The worker process fails. The app doesn't start.

> The app didn't reference the `Microsoft.AspNetCore.App` framework. Only apps targeting that framework can be hosted by ANCM.

Reviewer 2: ```suggestion
The app didn't reference the `Microsoft.AspNetCore.App` framework. Only apps targeting the `Microsoft.AspNetCore.App` framework can be hosted by the ASP.NET Core Module.
```

> To fix this error, make sure the app is targeting the `Microsoft.AspNetCore.App` framework. Check the `.runtimeconfig.json` to verify the framework targeted by the app.

Reviewer 2: ```suggestion
To fix this error, confirm that the app is targeting the `Microsoft.AspNetCore.App` framework. Check the `.runtimeconfig.json` to verify the framework targeted by the app.
```

### 500.34 ANCM Mixed Hosting Models Not Supported

> The worker process cannot run both an in-process and out-of-process application in the same worker process.

Reviewer 2: ```suggestion
The worker process can't run both an in-process app and an out-of-process app in the same process.
```

To fix this error, run apps in separate IIS application pools.

### 500.35 ANCM Multiple In-Process Applications in same Process

> The worker process cannot run two inprocess applications in the same worker process.

Reviewer 2: ```suggestion
The worker process can't run both an in-process app and an out-of-process app in the same process.
```

To fix this error, run apps in separate IIS application pools.

### 500.36 ANCM Out-Of-Process Handler Load Failure

> The out-of-process request handler, `aspnetcorev2_outofprocess.dll`, could not be found next to the `aspnetcorev2.dll`. This indicates a corrupted installation of the ASP.NET Core Module.

Reviewer 2: ```suggestion
The out-of-process request handler, *aspnetcorev2_outofprocess.dll*, isn't next to the *aspnetcorev2.dll* file. This indicates a corrupted installation of the ASP.NET Core Module.
```

> To fix this error, repair your installation of the Windows Hosting Bundle (for IIS) or Visual Studio (for IIS Express).

Reviewer 2: ```suggestion
To fix this error, repair the installation of the [.NET Core Hosting Bundle](xref:host-and-deploy/iis/index#install-the-net-core-hosting-bundle) (for IIS) or Visual Studio (for IIS Express).
```

### 500.37 ANCM Failed to Start Within Startup Time Limit

ANCM failed to start within the provied startup time limit. By default, the timeout is 120 seconds.

> This can occur when starting a large number of apps on the same machine. Check for CPU/Memory usage spikes on the server during startup. You may need to stagger the startup process of multiple apps.

Reviewer 2: ```suggestion
This error can occur when starting a large number of apps on the same machine. Check for CPU/Memory usage spikes on the server during startup. You may need to stagger the startup process of multiple apps.
```

### 500.30 In-Process Startup Failure

The worker process fails. The app doesn't start.

> The ASP.NET Core Module attempts to start the .NET Core CLR in-process, but it fails to start. The cause of a process startup failure can usually be determined from entries in the [Application Event Log](#application-event-log) and the [ASP.NET Core Module stdout log](#aspnet-core-module-stdout-log).

Reviewer 2: ```suggestion
The ASP.NET Core Module attempts to start the .NET Core runtime in-process, but it fails to start. The cause of a process startup failure is usually determined from entries in the [Application Event Log](#application-event-log) and the [ASP.NET Core Module stdout log](#aspnet-core-module-stdout-log).
```

### 500.0 In-Process Handler Load Failure

The worker process fails. The app doesn't start.

> An unknown error occurred loading ANCM components. Contact [Microsoft Support](https://support.microsoft.com/oas/default.aspx?prid=15832) (select "Developer Tools" then "ASP.NET Core") for assistance, ask on Stack Overflow, or file an issue on our [GitHub repository](https://github.com/aspnet/AspNetCore).

Reviewer 2: ```suggestion
An unknown error occurred loading ASP.NET Core Module components. Take one of the following actions:

* Contact [Microsoft Support](https://support.microsoft.com/oas/default.aspx?prid=15832) (select **Developer Tools** then **ASP.NET Core**).
* Ask a question on Stack Overflow.
* File an issue on our [GitHub repository](https://github.com/aspnet/AspNetCore).
```
